signature/cose/envelope.go

```diff
@@ -631,20 +635,61 @@ func decodeTime(timeRaw cbor.RawMessage) (time.Time, error) {
 	return t, nil
 }
 
-// parseTime parses time values from cose.ProtectedHeader
-// in go-cose, CBOR times (tag 1) decode to time.Time.
+// parseTime validates Tag1 Datetime in cborMap given label then returns
+// time.Time value from cose.ProtectedHeader.
 //
 // For more details: https://github.com/fxamacker/cbor/blob/7704fa5efaf3ef4ac35aff38f50f6ff567793072/decode.go#L52
-func parseTime(timeValue interface{}) (time.Time, error) {
-	switch t := timeValue.(type) {
+func parseTime(cborMap map[interface{}]cbor.RawMessage, label string, protected cose.ProtectedHeader) (time.Time, error) {
+	err := validateTag(cborMap, label)
+	if err != nil {
+		return time.Time{}, err
+	}
+
+	switch t := protected[label].(type) {
 	case cbor.RawMessage:
 		return decodeTime(t)
-	// TODO: need a way to check the tag number of datetime inside the signature
-	// and fail if it's a Tag0. We only accept Tag1 datetime.
-	// https://github.com/notaryproject/notation-core-go/issues/97
 	case time.Time:
 		return t, nil
 	}
 
 	return time.Time{}, errors.New("invalid timeValue type")
 }
+
+// generateRawProtectedCBORMap checks RawProtected of the COSE envelope Headers.
+// Returns the unmarshalled CBOR map of the RawProtected header.
+func generateRawProtectedCBORMap(rawProtected cbor.RawMessage) (map[interface{}]cbor.RawMessage, error) {
+	var cborMap map[interface{}]cbor.RawMessage
+	if len(rawProtected) == 0 {
+		return cborMap, nil
+	}
+
+	var decoded []byte
+	err := decMode.Unmarshal(rawProtected, &decoded)
+	if err != nil {
+		return cborMap, err
+	}
+	err = cbor.Unmarshal(decoded, &cborMap)
+	if err != nil {
+		return cborMap, err
+	}
+
+	return cborMap, nil
+}
+
+// validateTag validates the tag number of the input field. Returns error
+// if the tag number is not 1.
+func validateTag(cborMap map[interface{}]cbor.RawMessage, label interface{}) error {
+	rawMsg, ok := cborMap[label]
+	if ok {
+		rawTag := &cbor.RawTag{}
+		err := rawTag.UnmarshalCBOR([]byte(rawMsg))
+		if err != nil {
+			return err
+		}
+		if rawTag.Number != 1 {
+			return errors.New("only Tag1 Datetime CBOR object is supported")
+		}
+	}
+
+	return nil
+}
```

Comment:
I suggest merging this function to `parseTime` since `ok` is always `true`.

Reply:
Merged into parseTime. Added a check for `ok`, although theoretically, it should always be `true`.